src/workflow/utils.ts

```diff
@@ -43,6 +50,95 @@ export class WorkflowUtils {
     }
   }
 
+  static validateOutputs(workflow: Workflow) {
+    const { steps: workflowSteps } = workflow;
+    const seenSteps = new Set(); // Set to track executed steps
+
+    function validateOutputReferences(stepName: string, template?: string, parentName?: string) {
+      if (!template) {
+        return;
+      }
+      let match: RegExpExecArray | null;
+
+      // eslint-disable-next-line no-cond-assign
+      while ((match = regexWorkflowOutput.exec(template)) !== null) {
+        const workflowName = match[1];
+        if (parentName !== workflowName) {
+          throw new WorkflowCreationError(
+            `Invalid output reference: ${match[0]}, step is not a child of ${parentName}`,
+            workflow.name,
+            parentName,
+            stepName,
+          );
+        }
+        const outputName = `${workflowName}.${match[2]}`;
+        if (!seenSteps.has(outputName)) {
+          throw new WorkflowCreationError(
+            `Invalid output reference: ${match[0]}`,
+            workflow.name,
+            parentName,
+            stepName,
+          );
+        }
+      }
+
+      // eslint-disable-next-line no-cond-assign
+      while ((match = regexSimpleOutput.exec(template)) !== null) {
+        const outputStepName = match[1];
+        if (!seenSteps.has(outputStepName)) {
+          throw new WorkflowCreationError(
+            `Invalid output reference: ${match[0]}`,
+            workflow.name,
+            parentName ?? stepName,
+            parentName ? stepName : undefined,
+          );
+        }
+      }
+    }
```

Comment:
Can you give me a gist of this condition I am not clearly able to understand

Reply:
added comments